### DISCUSSION TITLE
Unable to setup Jamulus repository

### DISCUSSION THREAD

**Discussion by ksjdragon** [jamulussoftware/jamulus#3180] [Technical Support & Help] [answered]
When running sudo apt update, apt is unable to pull the InRelease file.

> Err:17 https://github.com/jamulussoftware/jamulus/releases/latest/download ./ InRelease
  401  Unauthorized [IP: 185.199.109.133 443]
Reading package lists... Done
E: Failed to fetch https://github.com/jamulussoftware/jamulus/releases/latest/download/./InRelease  401  Unauthorized [IP: 185.199.109.133 443]
E: The repository 'https://github.com/jamulussoftware/jamulus/releases/latest/download ./ InRelease' is not signed.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.

Of course, I can go to the URL there and download it without an issue. Any ideas as to what the issue could be?

**Comment by pljones** [comment]
Could you please provide a console log in full of what commands you've entered prior to the error.

**Reply by ksjdragon** [reply]
me@computer:$ curl https://raw.githubusercontent.com/jamulussoftware/jamulus/main/linux/setup_repo.sh > setup_repo.sh
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1110  100  1110    0     0   8538      0 --:--:-- --:--:-- --:--:--  8473
me@computer:$ chmod +x setup_repo.sh 
me@computer:$ sudo ./setup_repo.sh 
[sudo] password for me:                  
Setting up Jamulus repo at /etc/apt/sources.list.d/jamulus.list...
Installing Jamulus GPG key at /etc/apt/trusted.gpg.d/jamulus.asc...
Running apt update...
E: Failed to fetch https://github.com/jamulussoftware/jamulus/releases/latest/download/./InRelease  401  Unauthorized [IP: 185.199.111.133 443]
E: The repository 'https://github.com/jamulussoftware/jamulus/releases/latest/download ./ InRelease' is not signed.

**Reply by rdica** [reply]
Seems you have to add the `-L` argument to curl to download `jamulus.asc` key file.

**Reply by ksjdragon** [reply]
I have the key file downloaded, that's not the issue, It's present in /etc/apt/trusted.gpg.d/jamulus.asc

**Reply by rdica** [reply]
`apt-key list` shows it?

**Reply by ksjdragon** [reply]
```
/etc/apt/trusted.gpg.d/jamulus.asc
----------------------------------
pub   rsa4096 2023-07-27 [SC] [expires: 2026-07-26]
      DC5A A15A 46D0 90CA 884F  8143 B1EF 072B 7BD7 F31F
uid           [ unknown] The Jamulus Development Team (GPG key for signing Jamulus releases generated by the official deb repository) <team@jamulus.io>
sub   rsa4096 2023-07-27 [E] [expires: 2026-07-26]
```
Yes, it is present inside

**Reply by rdica** [reply]
```
root@daw8:~# curl https://raw.githubusercontent.com/jamulussoftware/jamulus/main/linux/setup_repo.sh > setup_repo.sh
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1110  100  1110    0     0   2573      0 --:--:-- --:--:-- --:--:--  2575
root@daw8:~# ll
total 44
drwx------  6 root root 4096 Sep 25 20:11 ./
drwxr-xr-x 19 root root 4096 Aug 14 14:52 ../
-rw-------  1 root root  635 Sep 23 16:17 .bash_history
-rw-r--r--  1 root root 3106 Oct 15  2021 .bashrc
drwx------  2 root root 4096 Aug 14 14:56 .cache/
drwxr-xr-x  3 root root 4096 Aug 15 16:21 .local/
-rw-r--r--  1 root root  161 Jul  9  2019 .profile
-rw-r--r--  1 root root 1110 Sep 25 20:11 setup_repo.sh
drwx------  3 root root 4096 Aug 14 14:53 snap/
drwx------  2 root root 4096 Aug 14 14:53 .ssh/
-rw-------  1 root root 1409 Sep 13 14:37 .viminfo
root@daw8:~# cat setup_repo.sh 
#!/bin/bash

# This script installs a Jamulus repository to Debian based systems

if [[ ${EUID} -ne 0 ]]; then
     echo "Error: This script must be run as root."
     exit 1
fi

REPO_FILE=/etc/apt/sources.list.d/jamulus.list
KEY_FILE=/etc/apt/trusted.gpg.d/jamulus.asc
GITHUB_REPOSITORY="jamulussoftware/jamulus"

echo "Setting up Jamulus repo at ${REPO_FILE}..."
echo "deb https://github.com/${GITHUB_REPOSITORY}/releases/latest/download/ ./" > ${REPO_FILE}
echo "Installing Jamulus GPG key at ${KEY_FILE}..."
curl --fail --show-error -sLo "${KEY_FILE}" https://github.com/${GITHUB_REPOSITORY}/releases/latest/download/key.asc

CURL_EXITCODE=$?
if [[ ${CURL_EXITCODE} -ne 0 ]]; then
      echo "Error: Download of gpg key failed. Please try again later."
      exit ${CURL_EXITCODE}
fi

echo "Running apt update..."
apt -qq update
echo "You should now be able to install a full Jamulus package via"
echo "  apt install jamulus"
echo "or a server-only, dependency-reduced build via"
echo "  apt install jamulus-headless"
echo
echo "This package will automatically be updated when you perform system updates."
root@daw8:~# chmod 700 setup_repo.sh 
root@daw8:~# ./setup_repo.sh 
Setting up Jamulus repo at /etc/apt/sources.list.d/jamulus.list...
Installing Jamulus GPG key at /etc/apt/trusted.gpg.d/jamulus.asc...
Running apt update...
All packages are up to date.
You should now be able to install a full Jamulus package via
  apt install jamulus
or a server-only, dependency-reduced build via
  apt install jamulus-headless

This package will automatically be updated when you perform system updates.
root@daw8:~# cat /etc/os-release 
PRETTY_NAME="Ubuntu 22.04.3 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.3 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy
root@daw8:~# apt search jamulus
Sorting... Done
Full Text Search... Done
jamulus/unknown 3.10.0 amd64
  Low latency Audio Server/Client

jamulus-headless/unknown 3.10.0 amd64
  Low latency Audio Server (headless)
```
Seems to work for me on Ubuntu Server 22.04 LTS.

Interesting, fails on Ubuntu Server 20.04...

**Reply by ksjdragon** [reply]
I was looking up a few things about this, can you check the apt version on those systems? For reference, mine is 2.4.10. 
I think it might be a bug in apt in how it handles the redirect... 
From here:
https://github.com/linux-surface/linux-surface/issues/625

**Reply by rdica** [reply]
Jammy 22.04 has `2.4.10` which works for me. Focal 20.04 has `2.0.9` and doesn't work.

**Reply by rdica** [reply]
Perhaps the older apt doesn't automatically follow the redirect I mentioned when using `curl -L`?

**Reply by ksjdragon** [reply]
Sorry, turns out I was on 2.4.9. I managed to update to 2.4.10, and seemed to fix the issue.... 
Do you think it'd be a good idea to mention this on the docs?

**Reply by rdica** [reply]
I don't see an update to 2.4.10 for Ubuntu 20.04.
How did you upgrade apt?

**Reply by ksjdragon** [reply]
I'm on linux mint, I guess I was on an older version of it, so i figured I might as well update it... 

I think I saw some ways to install a newer version of apt / an apt with fixes that people pulled
https://launchpad.net/~gpxbv/+archive/ubuntu/apt-urlfix

But I wasn't on ubuntu, so I didn't try it. (I probably could have, since Linux Mint is basically using the ubuntu repositories anyway, but I just didn't want to bother)

**Reply by rdica** [reply]
The expectation is one keeps their system up to date before installing new software.
I'm glad you got something working but it appears there's an issue with at least Ubuntu 20.04.
Of course a workaround is to manually download the pkg and install it from local disk but then no autoupdating with the rest of the system.

Bug report?

**Reply by ksjdragon** [reply, marked as answer]
Well, it's not like my system is that out-of-date or anything, it was just on the previous version.  I got the repository for the apt-urlfix from here, which was the bug report, seeming to correspond to the same issue with git.

https://bugs.launchpad.net/ubuntu/+source/apt/+bug/1950095

**Comment by pljones** [comment]
> I'm glad you got something working but it appears there's an issue with at least Ubuntu 20.04.

Yeah, we could probably do with a bug for that, if we say we support 20.04 still.

**Reply by rdica** [reply]
Shall I submit a bug report then?

**Reply by ann0see** [reply]
Best, add a warning to the website.

**Comment by houbsta** [comment]
Thanks for the links and information. I can indeed see that Ubuntu 20.04 has an issue with signed repositories with a certain type of URL with unencoded query strings.

I have tried : 

- Updating apt to use focal-proposed, which shunts the apt version to 2.0.10
- Setting `[trusted=yes]` in /etc/sources.list.d/jamulus.list
- Importing the keys manually with apt-key add

Since none of this worked, I cannot see a reasonable workaround for Ubunto 20.04, which means the repo is not supported under Ubuntu 20.04 (you already knew that, I'm just confirming).

So I'm updating to Ubuntu 22.04. For regular users installing the package manually under Ubuntu 20.04 works and is the best solution for them. The repo should be marked as unsupported. Indeed, you might even detect 20.04 in `setup-repo.sh` and state it isn't supported, since it will throw errors for every `apt update` and `apt upgrade` operation once installed, which adds to the confusion for the neophyte.

**Reply by ann0see** [reply]
> Indeed, you might even detect 20.04

Yes, I think that's a good solution. Can you open an issue for that please? I think it should be possible to detect the faulty apt version. 

> For regular users installing the package manually under Ubuntu 20.04 works and is the best solution for them. 

That's probably true, but if I were in their case, I'd just upgrade.
